### ISSUE TITLE
is fx803（rx580） support cudnn in ROCm6.2？

### ISSUE THREAD

**Issue by Kinghammer1** [lshqqytiger/ZLUDA#115]
is fx803（rx580） support cudnn in ROCm6.2？

**Comment by advanced-lvl-up** [comment]
no, probably not.  i haven't tried it myself but i would guess that it doesn't work.    the Windows HIP SDK 6.24  can work with these older archs though if you replace the nvcuda.dll file in your zluda directory with the one meant for HIP SDK 5.7.

Generally i think its recommended just to use ROCm HIP SDK 5.71.

There's now a fully open source fix for GFX803 & GFX900 from me that removes the crashing and memory leak issues.  I use it myself and its works fine.  Faster than a nvidia 3060 laptop for some basic diffusion tasks.   (power throttled laptop on battery LOL)
I'm in the process of cleaning up the build instructions and adding .yaml templates for the casual user who wants to build the fixed Tensile library and rocblas.dll for themselves.

For FLUX1 Schnell GGUF on a RX580 8GB the test system is about 21 seconds per iteration.  you can do stuff like video diffusion in the smaller wan 1.3b parameter model in 3 to 8 minutes.  and the full 14b model runs fairly well at 384x576 resolution in about 10 to 30 minutes per video also.  So dont let random redditors tell you these old cards don't work :)  img2vid might have some issues though since there are less memory optimizations available.

there's a comfyui zluda fork from patientx that i would recommend. or you can use the main fork and modify the 'memory_management.py' file yourself quite easily.  For manual installation install pytorch version 2.2.0 or something like that, and follow the usual zluda instructions.  Good luck.

**Comment by CS1o** [comment]
Hey maybe you can add your tensile files and rocblas.dll to the gfx collection here:
https://github.com/likelovewant/ROCmLibs-for-gfx1103-AMD780M-APU

**Comment by 1635200412** [comment]
It probably won't work. ROCm 6 requires amdhip64_6.dll, but the latest RX 580 driver doesn't even have this DLL! I tried forcing it to call this DLL, but it still couldn't recognize the RX 580!!!

**Comment by advanced-lvl-up** [comment]
@CS1o Right, i totally forgot about that.  thanks for your support and testing, guys.   Project isn't abandoned, actually made some progress this morning.   I've decided the tensile fix is just a niche thing for people who love old hardware, meaning i wont talk about it on popular media, etc.   But its also free for anyone to compile, copy , share, talk about, improve it, if they so choose.

cuDNN:  ok, at first i just assumed these archs had zero MIOpen support, but now looking at the HIP SDK extension files again, i noticed they are indeed in there.  Could be a relic from the original Linux fork or something?  I guess that means it's possible to implement.. or not?  We'll have to ask our resident expert here on how difficult partial support would be.

---------------------------------------------------------------------
(**Major Update, New Crash Fix.  the following is unrelated to cuDNN**): 

Speaking of which,  @lshqqytiger  once mentioned that the main issue with these unsupported architectures was a memory bug in the drivers.   With my custom modified comfyui install that i use personally, i never had this issue since it handles memory more directly than the main fork, hence the mystery of why people report crashes when i almost never see it.

So I had some time this morning and gave it some more thought, and here is the solution i came up with.  Basically, i think it's simply that the ROCm enumerator doesn't fully detect these unsupported archs.  Which is why some pytorch implementations try to allocate excessive 15GB+ VRAM loads occasionally in certain problematic VAE related workflows/nodes.

Fortunately the fix should be super easy, in theory.  Just force tiled VAE encode, and  add a manual check for wrong user settings (like trying to render 8k resolution videos, etc)  during the main workflow loop, or inside relevant nodes.  This should allow even older 4GB VRAM AMD cards to run demanding workloads that once failed.   Make a backup then edit the following file:

\your_path_to_comfyUI\comfy\sd.py   (same directory as model_management.py),  near line 624 in sd.py, of the latest commit:
inside "def encode(self, pixel_samples):"  You'll find the following code near line 630:

`        try:`
`            memory_used = self.memory_used_encode(pixel_samples.shape, self.vae_dtype)`
`            model_management.load_models_gpu([self.patcher], memory_required=memory_used, force_full_load=self.disable_offload)`
`            free_memory = model_management.get_free_memory(self.device)`
`            batch_number = int(free_memory / max(1, memory_used))`
`            batch_number = max(1, batch_number)`
`            samples = None`
`            for x in range(0, pixel_samples.shape[0], batch_number):`
`                pixels_in = self.process_input(pixel_samples[x:x + batch_number]).to(self.vae_dtype).to(self.device)`
`                out = self.first_stage_model.encode(pixels_in).to(self.output_device).float()`
`                if samples is None:`
`                    samples = torch.empty((pixel_samples.shape[0],) + tuple(out.shape[1:]), device=self.output_device)`
`                samples[x:x + batch_number] = out`

to fix crashing on cards with less 16GB VRAM (maybe even helps on 16GB VRAM Vega, untested),   you can change the previous code to the something like this:

`        try:`
`            memory_used = self.memory_used_encode(pixel_samples.shape, self.vae_dtype)`
`            model_management.load_models_gpu([self.patcher], memory_required=memory_used, force_full_load=self.disable_offload)`
`            free_memory = model_management.get_free_memory(self.device)`
`            batch_number = int(free_memory / max(1, memory_used))`
`            batch_number = max(1, batch_number)`
`            samples = None`
`            #Quick fix for AMD late GCN archs with less than 16GB VRAM`
`            logging.warning("Note: VAE encode allocation safe mode, attempting quick fix. Forcing tiled encode")`
`            if self.latent_dim == 3:`
`                tile = 256`
`                overlap = tile // 4`
`                samples = self.encode_tiled_3d(pixel_samples, tile_x=tile, tile_y=tile, overlap=(1, overlap, overlap))`
`            elif self.latent_dim == 1 or self.extra_1d_channel is not None:`
`                samples = self.encode_tiled_1d(pixel_samples)`
`            else:`
`                samples = self.encode_tiled_(pixel_samples)`

And it works!  on the very limited 8GB gfx803 test system, instead of crashing in certain IMG2VID first & last frame Wan nodes, it can now render native resolution (832x480) x 81 frames 5 seconds video, easily in about 20 minutes.  (using the toy 1.3b model)  You can use similar style code or adapt it for ForgeUI, etc, if you write it yourself.   I didn't bother writing any code at all (bounds checks, etc, for enabling normal faster VAE encode)  because i'm lazy.   

Test used the following:  Wan VACE 1.3b,  WanFunInpainting 1.3b, GGUF, Standard Wan model for VAE encoding, TinyVAE model for VAE decoding,  DMD self-forcing lora,   The above code forces tiled VAE ENCODE only,  TinyVAE reduces VRAM usage during DECODE, so thats two fixes.  For gfx803 & gfx900, i recommend almost always using TinyVAE decode.    The .pth model files from author madebyollin's excellent TinyVAE, goes into \yourpath_to_comfyui\models\vae_approx    -- it's the only ".pth" file i recommend, as more  generally you should only use gguf and safetensors from trusted sources.

comfy commandline args: "--disable-cuda --use-quad-cross-attention  --force-fp16 --fp16-unet --fp16-vae --novram"
some of these arguments maybe redundant / deprecated,  just what i happened to try.  Sometimes removing --force-fp16 and specifying no default dtypes  at all can help in certain models.  --novram is good for unsupported archs.
For 4GB cards,  you can lower the resolution a bit.  If you choose TinyVAE for decoding, you need a very minor code change in the TinyVAE comfy node to enable mixed precision VAE, since TinyVAE only supports fp32 for video, last time i checked anyway.  In other words dual VAEs.  connect normal VAE to the encode phase,   TinyVAE to the final decode. 

The following change is only needed if you you choose --fp16-vae or --bf16-vae, (often a good idea, it reduces VRAM usage during VAE encode).   your path to  blepping's comfy node implementation of madebyollin's excellent TinyVAE maybe different if you auto installed, instead of manually installing custom nodes.   Here's the fix,  near line 80 of the following file:

\yourpath_to_comfyUI\custom_nodes\ComfyUI-bleh-main\py\nodes\taevid.py

`        samples = latent["samples"].detach().to(device=device, dtype=dtype, copy=True)`

change that single line to this:

`        samples = latent["samples"].detach().to(device=device, dtype=torch.float, copy=True)`

Now you can used mixed precision dual VAEs.  That's all.  My uptime for the test system has been 8 months now,  rendering thousands of training data images and videos for prompt testing the other system.  With the above knowledge and my tensile fix, no crashes at all, 72 celsius.  Its a bit incomplete , some tensile solutions for productivity type apps are probably missing.  Will check when i have time.

I'll publish the full workflow and details soon, plus more AMD tips.   Keep an eye on my github page for more interesting fixes and workflows like this.  They sometimes help even on Nvidia or later gen AMD cards if you're experiencing BIOS compatibility or Dependency Hell issues. 

TLDR:  by using zluda, and the Tensile fix from my github page, plus that minor code change, you can now run something like patientx's comfy fork,  using completely unsupported AMD hardware like 803.  And with that you can test powerful state-of-the-art models  that are better than some commercial apps.  Most people don't think these cards even work on modern apps, therefore you can literally find these 8GB vidcards for free.
